tests/unit/s2n_tls13_state_machine_handshake_test.c

```diff
@@ -305,6 +307,86 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_connection_free(conn));
     }
 
+    /* Test: TLS 1.3 handshakes all follow CCS middlebox compatibility rules.
+     * RFC: https://tools.ietf.org/html/rfc8446#appendix-D.4 */
+    {
+        bool change_cipher_spec_found;
+        uint32_t handshake_type;
+        message_type_t *messages;
+
+        /* Test: "If not offering early data, the client sends a dummy change_cipher_spec record immediately before
+         *       its second flight. This may either be before its second ClientHello or before its encrypted handshake flight." */
+        for (size_t i = 0; i < valid_tls13_handshakes_size; i++) {
+            change_cipher_spec_found = false;
+            handshake_type = valid_tls13_handshakes[i];
+            messages = tls13_handshakes[handshake_type];
+
+            for (size_t j = 1; j < S2N_MAX_HANDSHAKE_LENGTH; j++) {
+
+                /* Ignore initial handshakes */
+                if (!IS_NEGOTIATED(handshake_type)) {
+                    continue;
+                }
+
+                /* Is it the second client flight?
+                 * Have we switched from the server sending to the client sending? */
+                if (tls13_state_machine[messages[j]].writer != 'C'
+                        || tls13_state_machine[messages[j - 1]].writer != 'S') {
+                    continue;
+                }
+
+                EXPECT_EQUAL(messages[j], CLIENT_CHANGE_CIPHER_SPEC);
+                EXPECT_EQUAL(tls13_state_machine[messages[j + 1]].writer, 'C');
+
+                /* CCS message found. We are done with this handshake. */
+                change_cipher_spec_found = true;
+                break;
+            }
+
+            if (IS_NEGOTIATED(handshake_type)) {
+                EXPECT_TRUE(change_cipher_spec_found);
+            }
+        }
+
+        /* Test: "If offering early data, the [change_cipher_spce] record is placed immediately after the first ClientHello."
+         *
+         * TODO: This can't be tested yet because S2N doesn't support early data. */
```

Comment:
We should probably track this in an issue, we could update https://github.com/awslabs/s2n/issues/967 to make sure that this TODO is addressed when early data is added.